src/AST.purs

```diff
@@ -484,7 +501,58 @@ data Type
 data AD
     = TList Type
     | TTuple (List Type)
-
+    | TTypeCons String (List Type)
+
+-- ADT Definition
+--
+-- The definition of a Type consists of the name of the type,
+-- followd by type variables, over which it is parametrized
+-- and a list of Data Constructors, each having a name,
+-- and a list of types, which are their parameters.
+data ADTDef
+  = ADTDef String (List TVar) (List (DataCons Type))
+
+derive instance eqADTDef :: Eq ADTDef
+
+instance showADTDef :: Show ADTDef where
+  show (ADTDef n vs cs) =
+    "data "
+    <> n
+    <> " "
+    <> intercalate " " (map show vs)
+    <> "\n  = "
+    <> intercalate "\n  | " (map show cs)
+
+
+-- | DataConstructor parameterized over its parameters,
+--   to use it for both, type definitions and data.
+
+data Associativity
+  = LEFTASSOC
+  | RIGHTASSOC
+  | ASSOC
+
+derive instance eqAssociativity :: Eq Associativity
+
+data DataCons param
```

Comment:
Cons could be confused with the list cons (:), like in ConsLit below. Please use Constr everywhere.